src/utils/transactions.ts

```diff
@@ -93,45 +93,42 @@ export const hasOnlyInputsWith = (inputs: Input[], addresses: Address[]): boolea
 export const hasOnlyOutputsWith = (outputs: Output[], addresses: Address[]): boolean =>
   outputs.every((o) => addresses.map((a) => a.hash).indexOf(o?.address ?? '') >= 0)
 
-export const getDirection = (tx: Transaction, address: AddressHash): TransactionDirection => {
-  const amount = calAmountDelta(tx, address)
-  const amountIsBigInt = typeof amount === 'bigint'
-  return amount && amountIsBigInt && amount < 0 ? 'out' : 'in'
-}
+export const getDirection = (tx: Transaction, address: AddressHash): TransactionDirection =>
+  calAmountDelta(tx, address) < 0 ? 'out' : 'in'
 
-export const calculateAmountSent = (tx: Transaction | UnconfirmedTransaction): bigint => {
-  const outputs = tx.outputs ?? []
-  const total = outputs.reduce((acc, output) => acc + BigInt(output.attoAlphAmount), BigInt(0))
+export const calculateUnconfirmedTxSentAmount = (tx: UnconfirmedTransaction, address: AddressHash): bigint => {
+  if (!tx.inputs || !tx.outputs) throw 'Missing transaction details'
 
-  if (isConsolidationTx(tx)) return total
+  const totalOutputAmount = tx.outputs.reduce((acc, output) => acc + BigInt(output.attoAlphAmount), BigInt(0))
 
-  const inputAddresses = tx.inputs ? uniq(tx.inputs.map((input) => input.address)) : []
-  const change = outputs.reduce(
-    (acc, output) => (inputAddresses.includes(output.address) ? acc + BigInt(output.attoAlphAmount) : acc),
+  if (isConsolidationTx(tx)) return totalOutputAmount
+
+  const totalOutputAmountOfAddress = tx.outputs.reduce(
+    (acc, output) => (output.address === address ? acc + BigInt(output.attoAlphAmount) : acc),
     BigInt(0)
   )
 
-  return total - change
+  return totalOutputAmount - totalOutputAmountOfAddress
```

Comment:
So, basically, since we cannot know the amounts of the input UTXOs, we cannot calculate the delta by substracting the sum of inputs from the sum of outputs of the specified address. Second best thing we can do is to substract the total outputs of the address (aka, change) from the total tx output amount.

This is pretty much what you did as well @LeeAlephium, right (reference to [Lee's code](https://github.com/alephium/desktop-wallet/pull/374/commits/f514957b15a9b9cfe906b16a56403703d5944ff5))? Please, let me know if you think this logic is missing something.

I believe that both of our approaches will produce wrong results in the special case of a transaction where there are more than 1 unique input address.

Reply:
If it appears to work then I guess it works :shrug: